builtin/providers/aws/resource_aws_elastic_beanstalk_application_version.go

```diff
@@ -0,0 +1,163 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/elasticbeanstalk"
+)
+
+func resourceAwsElasticBeanstalkApplicationVersion() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsElasticBeanstalkApplicationVersionCreate,
+		Read:   resourceAwsElasticBeanstalkApplicationVersionRead,
+		Update: resourceAwsElasticBeanstalkApplicationVersionUpdate,
+		Delete: resourceAwsElasticBeanstalkApplicationVersionDelete,
+
+		Schema: map[string]*schema.Schema{
+			"application": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsElasticBeanstalkApplicationVersionCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).elasticbeanstalkconn
+
+	application := d.Get("application").(string)
+	description := d.Get("description").(string)
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	name := d.Get("name").(string)
+
+	s3Location := elasticbeanstalk.S3Location{
+		S3Bucket: aws.String(bucket),
+		S3Key:    aws.String(key),
+	}
```

Comment:
If `bucket` is optional, what happens here if the User omits it from the config? I assume a failure in the `CreateApplicationVersion` call. We need to conditionally added this, if that is the case